### PR TITLE
Feature/e2e tests

### PR DESCRIPTION
**Pull request by abhishekmaha23**
**Changes**
- Added framework for E2E testing
- Reverted server fix
- Fixed bug in BTA flight number and travel class generation

**E2E testing**
The following use cases are tested!
- Case 1 - Comprehensive use-case with all types of data. Basically checking the flow and errors with normal data.
- Case 2 - Ideal use case. All people who flew are in the HR files. Complete Atmosfair coverage.
- Case 3 - Edge case. All people present in the BTA leg data. Reducing complication by having complete Atmosfair (as much as possible).
- Case 4 - Edge-case. Ideally distributed people data (minimal BTA). Empty Atmosfair data.
- Case 5 - Same as Case 1. Here, we can manually verify the contents of hr_bta_matches, to_atmosfair and hr_issue_list (with a second run on the same data).

**Usage**
In the current structure of the app, it is not possible to run multiple cases sequentially at the same time. Hence, each case must be run separately.
To create more cases, a config dict must be created as present in existing test cases in test_e2e.py. The following params are possible to be set.
- Number of total people (individuals from the leg data who may be either employees or guests
- Number of employees who have not taken flights.
- How people who have taken flights are distributed across the 3 classes (Employees, Guests, Unregistered employees)
- How the flight leg data sources are distributed across the 3 sources (Spesen records, Airplus records and BTA records)
- Proportion of records previously present in the input Atmosfair responses. (If 1, then no new records need to be requested)

Once the config is set
- reset_project() clears the data-tests, output-tests and data-cached folders
- mock_data_generator(config) copies the template files over to the empty data-tests folder, and generates all the relevant data
- the run_case() methods at the end of the file run the relevant case. 

**Running the Tests**
The TEST variable has to be set at least once through the following commands
- To run E2E tests, the following commands will work
--- In Windows
--- $env:TEST="E2E";docker-compose up --build
--- In Linux/Mac
--- TEST=E2E docker-compose up --build
- To run the normal app, the value should be set to an empty string
--- In Windows
--- $env:TEST="";docker-compose up --build
--- In Linux/Mac
--- TEST= docker-compose up --build

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,8 +4,8 @@ FROM amancevice/pandas:1.2.0-slim
 
 # for string_grouper
 RUN DEBIAN_FRONTEND=noninteractive apt-get update \
-&& apt-get install g++ -y \
-&& apt-get clean
+    && apt-get install g++ -y \
```
Comment (bebbi) [inline Dockerfile:7]: There might be a few formatting changes such as this. For this, and also minor style nits such as trailing whitespace or missing newlines at end of file: Can you check that these are not getting into commits, or are intentional improvements? I use autopep8 for formatting files before save, should we standardize on this?

File: app.py [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 import datetime as dt
+from config import config
 from atmosfair_test import atmosfair_test
 from anonymize_dataset import anonymize_dataset, move_dates
 from airports import get_legs_airport_data
 from hr import assign_employee_id8, find_hr_matches, hr_bta_fill_instructions, \
     demographics_by_year, merge_demographics, get_org_fte, hr_issues
 from bta_legs_import import bta_legs_import
 from regular_legs_import import regular_legs_import
-from config import config
```
Comment (bebbi) [inline app.py:10]: Like above - it is usually best to run a quick git diff before committing so that there are no "noise" changes in the commit.

**Review by bebbi — CHANGES_REQUESTED** [review]
The 2 comments I've added are minor. However, I'm still not able to run. Does the docker-compose command work for you? Here I had 2 issues:
- having to create data-tests manually
- Then, error FileNotFoundError: [Errno 2] No such file or directory: 'data-tests/airports-extended.csv'
Should this be fixed or is this an error on my side?

**Comment by bebbi** [comment]
As an additional comment, would you be willing to check/amend the VS Code .vscode config so I can run in VS Code debugger? If not, I can do that.

File: Dockerfile (hunk 1 above)
Comment (abhishekmaha23) [reply Dockerfile:7]: This whitespace was unintended, I'll change this!

File: app.py (hunk 2 above)
Comment (abhishekmaha23) [reply app.py:10]: This wasn't meant to be a noise change! The code imports the config yaml at this point after running all the other code, and I wanted to make sure that it was imported first

**Comment by abhishekmaha23** [comment]
> The 2 comments I've added are minor. However, I'm still not able to run. Does the docker-compose command work for you? Here I had 2 issues:
> 
> * having to create data-tests manually
> * Then, error FileNotFoundError: [Errno 2] No such file or directory: 'data-tests/airports-extended.csv'
>   Should this be fixed or is this an error on my side?

- Regarding the data-tests file - I was following the same format, since data-originals also must be manually created! But I can modify this so that the code creates a folder as such
- Actually I wanted to handle a case in the restructuring where the code doesn't run without the airports-extended file being defined, or having a value! For now it should work with just placing the airports-extended.csv from the data-originals within the test-base-files folder

**Comment by abhishekmaha23** [comment]
> As an additional comment, would you be willing to check/amend the VS Code .vscode config so I can run in VS Code debugger? If not, I can do that.

I'm actually not sure how to do this - I'll try this out

**Comment by abhishekmaha23** [comment]
I've updated the tasks.json in the last commit to do the env variable mapping and volume mapping for the newly built images.
The debugging functionality should hopefully work as expected now.

**Comment by bebbi** [comment]
Thanks a lot for your work to fix this. I added another commit to re-instate my changes to the vscode service.
Specifically, it's to enable 2 configurations, one for regular run, one for e2e run.
I noticed your addition of volumes in the .vscode config, where now it makes sense that code is not copied inside the Dockerfile any longer.

**Pull request merged**